document/core/valid/types.rst

```diff
@@ -272,6 +261,227 @@ Function Types
    }
 
 
+.. index:: compound type, function type, aggregate type, structure type, array type, field type
+   pair: validation; compound type
+   pair: validation; aggregate type
+   pair: validation; structure type
+   pair: validation; array type
+   single: abstract syntax; compound type
+   single: abstract syntax; function type
+   single: abstract syntax; structure type
+   single: abstract syntax; array type
+   single: abstract syntax; field type
+.. _valid-comptype:
+.. _valid-aggrtype:
+.. _valid-structtype:
+.. _valid-arraytype:
+
+Compound Types
+~~~~~~~~~~~~~~
+
+:math:`\TFUNC~\functype`
+........................
+
+* The :ref:`function type <syntax-functype>` :math:`\functype` must be :ref:`valid <valid-functype>`.
+
+* Then the compound type is valid.
+
+.. math::
+   \frac{
+     C \vdashfunctype \functype \ok
+   }{
+     C \vdashcomptype \TFUNC~\functype \ok
+   }
+
+:math:`\TSTRUCT~\fieldtype^\ast`
+................................
+
+* For each :ref:`field type <syntax-fieldtype>` :math:`\fieldtype_i` in :math:`\fieldtype^\ast`:
+
+  * The :ref:`field type <syntax-fieldtype>` :math:`\fieldtype_i` must be :ref:`valid <valid-fieldtype>`.
+
+* Then the compound type is valid.
+
+.. math::
+   \frac{
+     (C \vdashfieldtype \X{ft} \ok)^\ast
+   }{
+     C \vdashcomptype \TSTRUCT~\X{ft}^\ast \ok
+   }
+
+:math:`TARRAY~\fieldtype`
+.........................
+
+* The :ref:`field type <syntax-fieldtype>` :math:`\fieldtype` must be :ref:`valid <valid-fieldtype>`.
+
+* Then the compound type is valid.
+
+.. math::
+   \frac{
+     C \vdashfieldtype \X{ft} \ok
+   }{
+     C \vdashcomptype \TARRAY~\X{ft} \ok
+   }
+
+
+.. index:: field type, storage type, packed type, value type, mutability
+   pair: validation; field type
+   pair: validation; storage type
+   pair: validation; packed type
+   single: abstract syntax; field type
+   single: abstract syntax; storage type
+   single: abstract syntax; packed type
+   single: abstract syntax; value type
+.. _valid-fieldtype:
+.. _valid-storagetype:
+.. _valid-packedtype:
+
+Field Types
+~~~~~~~~~~~
+
+:math:`\mut~\storagetype`
+.........................
+
+* The :ref:`storage type <syntax-storagetype>` :math:`\storagetype` must be :ref:`valid <valid-storagetype>`.
+
+* Then the field type is valid.
+
+.. math::
+   \frac{
+     C \vdashstoragetype \X{st} \ok
+   }{
+     C \vdashfieldtype \mut~\X{st} \ok
+   }
+
+:math:`\packedtype`
+...................
+
+* The packed type is valid.
+
+.. math::
+   \frac{
+   }{
+     C \vdashpackedtype \packedtype \ok
+   }
+
+
+.. index:: recursive type, sub type, compound type, final, subtyping
+   pair: abstract syntax; recursive type
+   pair: abstract syntax; sub type
+.. _valid-rectype:
+.. _valid-subtype:
+
+Recursive Types
+~~~~~~~~~~~~~~~
+
+:ref:`Recursive types <syntax-rectype>` are validated for a specific :ref:`type index <syntax-typeidx>` that denotes the index of the type defined by the recursive group.
```

Comment:
What about `C |-_{x} rec subtype* ok` (i.e. a subscript)?

I'm personally quite negative about `ok (x)` because it looks unstructured/like a typo, and we don't normally use brackets this way.

Reply:
Subscripts on turnstiles are usually used for distinguishing different relations or modes of relations. So I wouldn't want to put it there.

Not sure I follow what's so bad about `ok(x)` as a classifier, it's like an indexed ok. Note that I deliberately didn't put a space where you put it. :)

Reply:
Oh, I get it, I just noticed that the macros mess up the spacing in the rendered version. That should be fixed now.

Reply:
I think you pushed this on the other PR's branch

Reply:
Oops, yes, pushed here now.